keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -300,10 +300,12 @@ _For example, [CAPN's virtualcluster project](https://github.com/kubernetes-sigs
 The `ClusterProperty` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterProperty`s, each holding a different identification related value. Each property contains the following information:
 
 *   **Name** - a well known or custom name to identify the property.
-*   **Value** - a property-dependent string, up to 128 KB.
+*   **Value** - a property-dependent string, up 128k Unicode code points (see _Note_).
```

Comment:
Uggh I hate this topic.  I can't keep it straight in my head.

If we are asserting length in OpenAPI, do we need to assert an ecoding (e.g. UTF8)?  What if the value is NOT valid unicode - what will the validator do?   E.g. can I craft a value of entirely invalid bytes, and will that trip up the validator? 

Sorry to keep poking this, but if we get it wrong it's a trap for consumers.

IMO: 

If it's a string we should validate that it is actually all unicode characters in some encoding, of some max length in code points.

If it's an array of bytes we should validate the length in bytes and nothing more.

Thoughts?

Reply:
**tl;dr I think we leave the wording as is. UTF-32 is weird but I don't think we should solve that here**

Ok tested this empirically on a `kind` cluster using a CRD with `ClusterProperty.value`'s maxLength set to `3` (see [`describe crd`](https://gist.github.com/lauralorenz/0b3423366f8df1cabf29d348f44bd084#file-crd-out)), and trying to insert one CR whose `value` uses 2 code points ("🇺🇸"), one CR whose `value` uses 4 code points ("🇺🇸🇺🇸"), and also trying  with my YAML encoded in UTF-8, UTF-16, and UTF-32 ([methodology](https://gist.github.com/lauralorenz/0b3423366f8df1cabf29d348f44bd084#file-encoding-py)).

> If we are asserting length in OpenAPI, do we need to assert an ecoding (e.g. UTF8)?

Empirically I confirmed that OpenAPIv3 maxLength is based on the number of Unicode code points of the decoded string because "🇺🇸" passes validation while "🇺🇸🇺🇸" does not [1]. This is regardless of if my YAML was originally in UTF-8 or UTF-16 (and FYI "🇺🇸"  uses 8 and 10 bytes in UTF-8 and UTF-16 respectively). 

Watching from the outside my best guess is by the time OpenAPIv3's maxLength validation code gets involved, the input is already decoded, though there is a way to [specify encoding validation in OpenAPIv3](https://swagger.io/specification/#encoding-object) BUT it is not exposed in the k8s CRD version of OpenAPIv3 ([ref](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1#JSONSchemaProps)) and I also empirically observed it was not allowed when I tried anyways. I don't think we need to enforce an encoding since ultimately we are limiting based on code points -- as long as someone's bytes can get decoded into Unicode code points by the time the validator checks, OpenAPIv3 maxLength doesn't care about the encoding and neither do I.

Empirically I also observed using UTF-8 OR UTF-16 encoded YAML both work fine, but something doesn't seem to be able to decode UTF-32 as I cannot apply my YAML if it's in UTF-32 [1]. Assuming my methodology to get to UTF-32 encoded YAML file was all correct, then from the error thrown it seems like the yaml->json converter library used by kubectl maybe can't sniff that my file was UTF-32. I'm not sure k8s generally really cares if people have their YAML in UTF-32 but if they do this might be worth investigating more outside of the context of this KEP.

[1] See the [output](https://gist.github.com/lauralorenz/0b3423366f8df1cabf29d348f44bd084#file-samples-out).

> What if the value is NOT valid unicode - what will the validator do

I observe that if I put an invalid UTF-8 byte sequence in as the `value` aka something that cannot be interpreted into a Unicode code point when decoding from UTF-8, it gets validated against number of resulting replacement characters ("�") which is one per invalid byte. I think for our purposes this means it didn't blow up and that is acceptable (as I'm sure we don't intend to support invalid bytes in UTF-* for this field). Specifically I tried cases 3.1.3 and 3.5.1 in [this UTF-8 test file](https://www.w3.org/2001/06/utf-8-wrong/UTF-8-test.html). 

Importantly regarding the methodology of this test, I did not copy paste between browsers / text editors but instead painstakingly (though probably not in the most efficient way) manipulated it in a Python3 interpreter (which is the language I know the best when it comes to the quirks of the string implementation). In summary I `wget`ted that file, read that file into a Python interpreter as bytes only, concatenated the bad bytes into the `value` field of the bytes of my YAML, and wrote it to file as bytes and never decoded / reencoded it, and even double checked by reading it in as bytes again that the byte sequence was not the byte sequence for the replacement character -- in other words, I feel very confident the bad bytes were not turned into "�" until k8s did so sometime during `kubectl apply`.

> IMO:

> If it's a string we should validate that it is actually all unicode characters in some encoding, of some max length in code points.

It is a string, agree on max length is against code points, and for reasons above I don't think we should enforce an encoding and also we can't easily.

> If it's an array of bytes we should validate the length in bytes and nothing more.

It's not. The only openings this way seem to be [OpenAPIv3 formats](https://swagger.io/specification/#data-types) "byte" (base64 encoded) or "binary" (currently not supported by the k8s CRD implementation of OpenAPIv3), and which would require everyone to write CRs in a non-human readable way like this which we would never do:

```
apiVersion: about.x-k8s.io/v1alpha1
kind: ClusterProperty
metadata:
  name: clusterproperty-fits
spec:
  value: "8J+HuvCfh7g=" # base64(flag emoji in UTF-8)
```